client/client.go

```diff
@@ -1484,166 +1572,206 @@ func (c *Client) runHandleInvocation(msg *wamp.Invocation) {
-		case <-ctx.Done():
-			// Received an INTERRUPT message from the router.
-			// Note: handler is also just as likely to return on INTERRUPT.
-			result = InvokeResult{Err: wamp.ErrCanceled}
-			var reason string
-			if errors.Is(ctx.Err(), context.DeadlineExceeded) {
-				reason = "callee due to timeout"
-			} else {
-				reason = "router"
-			}
-			c.log.Println("INVOCATION", reqID, "canceled by", reason)
+		// Create a kill switch so that invocation can be canceled.
+		if timeout > 0 {
+			// The caller specified a timeout, in milliseconds.
+			ctx, cancel = context.WithTimeout(context.Background(),
+				time.Millisecond*time.Duration(timeout))
+		} else {
+			ctx, cancel = context.WithCancel(context.Background())
 		}
-
-		if result.Err != "" {
-			c.sess.SendCtx(c.ctx, &wamp.Error{
-				Type:        wamp.INVOCATION,
-				Request:     reqID,
-				Details:     wamp.Dict{},
-				Arguments:   result.Args,
-				ArgumentsKw: result.Kwargs,
-				Error:       result.Err,
-			})
-			return
+		c.invHandlersCtxs[cliInvocation] = ctx
+		c.invHandlerKill[reqID] = cancel
+		c.activeInvHandlers.Add(1)
+
+		// If caller is accepting progressive results, create map entry to allow
+		// progress to be sent.
+		if progResOK {
+			c.progGate[ctx] = reqID
 		}
+	}
 
-		options := result.Options
-
-		if options == nil {
-			options = wamp.Dict{}
-		}
+	handlerQueue <- msg
 
-		message := &wamp.Yield{
-			Request: reqID,
-			Options: options,
-		}
+	c.sess.Unlock()
 
-		// Check and handle Payload PassThru Mode
-		// @see https://wamp-proto.org/wamp_latest_ietf.html#name-payload-passthru-mode
-		if pptScheme, _ := options[wamp.OptPPTScheme].(string); pptScheme != "" {
-			// Let's check: was ppt feature announced by dealer?
-			if !c.sess.HasFeature(wamp.RoleDealer, wamp.FeaturePayloadPassthruMode) {
-				// It's protocol violation, so we need to abort connection
-				abortMsg := wamp.Abort{
-					Reason: wamp.ErrProtocolViolation,
-					Details: wamp.Dict{
-						wamp.OptError:   ErrPPTNotSupportedByRouter.Error(),
-						wamp.OptMessage: ErrPPTNotSupportedByPeer.Error(),
-					},
+	if !queueExists {
+		// Start a goroutine to run the user-defined invocation handler.
+		go func() {
+			defer cancel()
+
+			// Create channel to hold result.  Channel must be buffered.
+			// Otherwise, canceling the call will leak the goroutine that is
+			// blocked forever waiting to send the result to the channel.
+			resChan := make(chan InvokeResult, 1)
+			go func() {
+				for msg := range handlerQueue {
+
+					if isInProgress, _ := msg.Details[wamp.OptProgress].(bool); !isInProgress {
+						c.sess.Lock()
+						close(c.invHandlersQueues[cliInvocation])
+						delete(c.invHandlersQueues, cliInvocation)
+						delete(c.invHandlersCtxs, cliInvocation)
+						c.sess.Unlock()
+					}
+
+					// The Context is passed into the handler to tell the client
+					// application to stop whatever it is doing if it cares to pay
+					// attention.
+					resChan <- handler(ctx, msg)
```

Comment:
I think this is the right place to create a WithTimeout context if the timeouts are per invocation